baseplate_test.go

```diff
@@ -145,6 +150,64 @@ func TestServe(t *testing.T) {
 	}
 }
 
+type timestampCloser struct {
+	ts []time.Time
+}
+
+func (c *timestampCloser) Close() error {
+	c.ts = append(c.ts, time.Now())
+	return nil
+}
+
+func TestServeClosers(t *testing.T) {
+	t.Parallel()
+
+	store := newSecretsStore(t)
+	defer store.Close()
+
+	bp := baseplate.NewTestBaseplate(baseplate.Config{StopTimeout: testTimeout}, store)
+
+	pre := &timestampCloser{}
+	post := &timestampCloser{}
+
+	args := baseplate.ServeArgs{
+		Server:       newWaitServer(t, bp, time.Millisecond),
+		PreShutdown:  []io.Closer{pre},
+		PostShutdown: []io.Closer{post},
+	}
+
+	ch := make(chan error)
+
+	go func() {
+		// Run Serve in a goroutine since it is blocking
+		ch <- baseplate.Serve(context.Background(), args)
+	}()
+
+	time.Sleep(time.Millisecond)
+	p, err := os.FindProcess(syscall.Getpid())
+	if err != nil {
+		t.Fatal(err)
```

Comment:
If we failed to get the pid, the already started goroutine will just be leaked (until all tests are done). I'd prefer to get pid first, fatal if failed, then start goroutine, sleep, and signal.